### DISCUSSION TITLE
Coupling kinetic Monte Carlo with Phase Field Model

### DISCUSSION THREAD

**Discussion by vermaprk** [idaholab/moose#23937] [Q&A Modules: Phase field] [answered]
Hi
I am trying to couple a simple kinetic Monte Carlo (kMC) problem of 2D random walk with the phase field model.
I previously posted regarding how to model a 2d random walk problem in moose.
https://github.com/idaholab/moose/discussions/23886

I want to implement the simplest code for 2d random walk in C as shown below: The system is 200 by 200 matrix with 40 particles initialized randomly. The movement of particles is in 4 directions (determined by the random number).
```
#include <stdio.h>
#include <random>

struct points{
  int x;
  int y;
};

int main(){
srand(time(NULL));
double rnum;
int np = 40;
int nwalk = 100;
int xmax = 200;
int ymax = 200;
struct points s[np];
struct points sfinal[nwalk][np];

for (int i=0;i<np;i++){
  int tempx = (rand() % (xmax + 1));
  int tempy = (rand() % (ymax + 1));
  s[i].x = tempx;
  s[i].y = tempy;
}

for (int i=0;i<nwalk;i++){

  for (int j=0;j<np;j++){
    rnum = rand() / (double)RAND_MAX ;
    if (rnum >=0 && rnum<=0.25 && s[j].x<xmax){
      s[j].x = s[j].x + 1;
    }      
    else if (rnum > 0.25 && rnum <=0.5 && s[j].x>0){
        s[j].x = s[j].x - 1;
    }
    else if (rnum > 0.5 && rnum <=0.75 && s[j].y<ymax){
        s[j].y = s[j].y + 1;
    }
    else if (rnum > 0.75 && rnum <=1 && s[j].y>0){ 
        s[j].y = s[j].y - 1;}

    sfinal[i][j].x = s[j].x;
    sfinal[i][j].y = s[j].y;
  }  
}
for (int i=0;i<nwalk;i++){
  printf("step %d\n",i);
  for(int j=0;j<np;j++){
    printf("particle %d = %d  %d\n",j,sfinal[i][j].x,sfinal[i][j].y);
  }
}
}
```

The output is:
[myVideo](https://user-images.githubusercontent.com/69255835/229488062-51299140-91d4-4c7c-96a4-e2eb77e7ad73.gif)

I tried the `raytracing` module. But, I am confused how to change the direction of rays randomly and to mimic the 2d random walk.
Is it possible to represent discrete particles using the raytracing kernels.
Please add any other methods if available to model this.

```
[Mesh]
  [gmg]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 10
    ny = 10
    xmax = 5
    ymax = 5
  []
[]

[Variables/u]
[]

[Kernels]
  [reaction]
    type = Reaction
    variable = u
  []
  [diffusion]
    type = Diffusion
    variable = u
  []
[]

[Executioner]
  type = Transient
  solve_type = PJFNK
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
  dt=2
  end_time = 50
  nl_abs_tol = 1e-10
[]

[UserObjects/study]
  type = ConeRayStudy

  start_points = '1 1.5 0'
  directions = '2 1 0'
  half_cone_angles = 2.5
  ray_data_name = weight

  # Must be set with RayKernels that
  # contribute to the residual
  execute_on = PRE_KERNELS

  # For outputting Rays
  always_cache_traces = true
[]

[RayBCs]
  [reflect]
    type = ReflectRayBC
    boundary = 'right top bottom'
  []
  [kill_rest]
    type = KillRayBC
    boundary = 'left'
  []
[]

[RayKernels/line_source]
  type = LineSourceRayKernel
  variable = u

  # Scale by the weights in the ConeRayStudy
  ray_data_factor_names = weight
[]

[Outputs]
  exodus = true
[] 
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

You should use the RayTracingExodus output and the Exodus output to look at what you are currently simulating. 

What you ll want to add is a RayKernel that simulates the change of directions. 
What dictates a change of direction of the particle? Just at random times?

Guillaume

**Reply by vermaprk** [reply]
Yes, right now the particle's direction changes randomly at each time step. Which raykernel can be used to depict such changes ?

**Reply by GiudGiud** [reply]
There is not an example, but you can call changeDirection inside a RayKernel on the particle/ray
